source/common/network/connection_impl.cc

```diff
@@ -731,19 +732,22 @@ void ConnectionImpl::flushWriteBuffer() {
 }
 
 ServerConnectionImpl::ServerConnectionImpl(Event::Dispatcher& dispatcher,
+                                           Server::ThreadLocalOverloadState& overload_state,
                                            ConnectionSocketPtr&& socket,
                                            TransportSocketPtr&& transport_socket,
                                            StreamInfo::StreamInfo& stream_info, bool connected)
     : ConnectionImpl(dispatcher, std::move(socket), std::move(transport_socket), stream_info,
-                     connected) {}
+                     connected),
+      overload_state_(overload_state) {}
 
 void ServerConnectionImpl::setTransportSocketConnectTimeout(std::chrono::milliseconds timeout) {
   if (!transport_connect_pending_) {
     return;
   }
   if (transport_socket_connect_timer_ == nullptr) {
     transport_socket_connect_timer_ =
-        dispatcher_.createTimer([this] { onTransportSocketConnectTimeout(); });
+        overload_state_.createScaledTimer(Server::OverloadTimerType::TransportSocketConnectTimeout,
```

Comment:
I did a local change that undid changes to this line and fixed all the unused variable compile errors, and noticed that the only test that failed //test/common/network:connection_impl_test due to changes to mock expectations.  It would be good to have some e2e integration test coverage for this functionality.

This issue also came up in https://github.com/envoyproxy/envoy/pull/14155 .  We have multiple scaled timers, we should have some generic recipes for how to integration test scaled timer functionality.  Possible strawman drive the connection to a certain state, force proxy into overload and verify that the timeout triggers shortly afterwards.

Reply:
Sent https://github.com/envoyproxy/envoy/pull/14290 to do just that for the existing scaled timer.